src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -8140,130 +8154,198 @@ void checkWellKnown(WellKnownMember member)
             // 2. For Index: Has an accessible indexer with a single int parameter
             //    For Range: Has an accessible Slice method that takes two int parameters
 
-            var lookupResult = LookupResult.GetInstance();
-
-            if (TryLookupLengthOrCount(receiverType, lookupResult, out lengthOrCountProperty, ref useSiteInfo) &&
-                TryFindIndexOrRangeIndexerPattern(lookupResult, receiverOpt, receiverType, argIsIndex, out patternSymbol, diagnostics, ref useSiteInfo))
+            if (TryBindLengthOrCount(syntax, receiverPlaceholder, out lengthOrCountAccess, diagnostics) &&
+                tryBindIndexOrRangeImplicitIndexer(syntax, receiver, argIsIndex, out implicitIndexerAccess, out argumentPlaceholders, diagnostics))
             {
-                CheckImplicitThisCopyInReadOnlyMember(receiverOpt, lengthOrCountProperty.GetMethod, diagnostics);
-                lookupResult.Free();
                 return true;
             }
 
-            patternSymbol = null;
-            lookupResult.Free();
+            lengthOrCountAccess = null;
+            implicitIndexerAccess = null;
+            argumentPlaceholders = default;
             return false;
-        }
 
-        private bool TryFindIndexOrRangeIndexerPattern(
-            LookupResult lookupResult,
-            BoundExpression? receiverOpt,
-            TypeSymbol receiverType,
-            bool argIsIndex,
-            [NotNullWhen(true)] out Symbol? patternSymbol,
-            // PROTOTYPE(list-patterns) We should take either of these and adjust the caller.
-            BindingDiagnosticBag diagnostics,
-            ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
-        {
-            Debug.Assert(lookupResult.IsClear);
-            if (argIsIndex)
+            // Binds pattern-based implicit indexer:
+            // - for Index indexer, this will find `this[int]`.
+            // - for Range indexer, this will find `Slice(int, int)` or `string.Substring(int, int)`.
+            bool tryBindIndexOrRangeImplicitIndexer(
```

Comment:
>tryBindIndexOrRangeImplicitIndexer [](http://example.com/codeflow?start=17&length=34)

As a general comment. Moving functions around just to make them local significantly complicates the review process. Effectively, this makes it impossible for reviewers to "build" on top of previous result.